### PR TITLE
AnnotationTypeMappings does not filter repeatable annotations

### PR DESCRIPTION
**Pull request by yilianhuaixiao**
since
`					if (!isMappable(source, metaAnnotation)) {
						continue;
					}`
is called before and 'source' ,'metaAnnotation' did not changed. 
The code should be deleted.

### CODE REVIEW

File: spring-core/src/main/java/org/springframework/core/annotation/AnnotationTypeMappings.java [hunk 1]
```diff
@@ -91,9 +91,6 @@ private void addMetaAnnotationsToQueue(Deque<AnnotationTypeMapping> queue, Annot
 					.findRepeatedAnnotations(metaAnnotation);
 			if (repeatedAnnotations != null) {
 				for (Annotation repeatedAnnotation : repeatedAnnotations) {
-					if (!isMappable(source, metaAnnotation)) {
```
Comment (sbrannen) [inline spring-core/src/main/java/org/springframework/core/annotation/AnnotationTypeMappings.java:94]: I'm wondering if that was perhaps a copy-and-paste error and if it should rather be:

```java
if (!isMappable(source, repeatedAnnotation)) {
    continue;
}
```

@philwebb, what are your thoughts on the matter?

**Comment by yilianhuaixiao** [comment]
@sbrannen  Yes, I think you are right

**Comment by sbrannen** [comment]
@yilianhuaixiao with your latest commit that _fixes_ the issue, please introduce a test that fails before the fix and passes after the fix.

**Comment by yilianhuaixiao** [comment]
@sbrannen  I added the testcase which can be passed after changing.

File: spring-core/src/main/java/org/springframework/core/annotation/AnnotationTypeMappings.java (hunk 1 above)
Comment (sbrannen) [reply spring-core/src/main/java/org/springframework/core/annotation/AnnotationTypeMappings.java:94]: @philwebb, I'm assuming my above assumption is correct. (I know: a lot of assuming).

In light of that, we'll go forward with the change in this PR.

File: spring-core/src/test/java/org/springframework/core/annotation/AnnotationTypeMappingsTests.java [hunk 2]
```diff
@@ -474,6 +474,16 @@ void isEquivalentToDefaultValueWhenNotMatchingReturnsFalse() {
 		return result;
 	}
 
+	@Test
+	void forAnnotationTypeWhenRepeatableMetaAnnotationFilterd() {
+		AnnotationTypeMappings mappings = AnnotationTypeMappings.forAnnotationType(WithRepeatedMetaAnnotations.class,
+		annotationType ->
+				ObjectUtils.nullSafeEquals(annotationType, Repeating.class.getName()));
```
Comment (sbrannen) [inline spring-core/src/test/java/org/springframework/core/annotation/AnnotationTypeMappingsTests.java:481]: You omitted the import for `ObjectUtils`, but don't worry about it. I'll add it before merging.

Comment (sbrannen) [reply spring-core/src/test/java/org/springframework/core/annotation/AnnotationTypeMappingsTests.java:481]: For future reference, I was able to simplify the annotation filter in that test method as follows.

```java
@Test
void forAnnotationTypeWhenRepeatableMetaAnnotationIsFiltered() {
	AnnotationTypeMappings mappings = AnnotationTypeMappings.forAnnotationType(WithRepeatedMetaAnnotations.class,
			Repeating.class.getName()::equals);
	assertThat(getAll(mappings)).flatExtracting(AnnotationTypeMapping::getAnnotationType)
			.containsExactly(WithRepeatedMetaAnnotations.class);
}
```

However, I won't be merging that change due to reasons I'll explain later.

**Comment by sbrannen** [comment]
@yilianhuaixiao, thanks again for the PR and the test case!

It turns out that `AnnotationTypeMappings.forAnnotationType(Class<? extends Annotation>, AnnotationFilter)` is not invoked outside of the `AnnotationTypeMappings` class itself.

**Comment by sbrannen** [comment]
@yilianhuaixiao, thanks again for the PR and the test case!

It turns out that `AnnotationTypeMappings.forAnnotationType(Class<? extends Annotation>, AnnotationFilter)` is not invoked outside of the `AnnotationTypeMappings` class itself.

It is only invoked from `AnnotationTypeMappings.forAnnotationType(Class<? extends Annotation>)`, with `AnnotationFilter.PLAIN` as the supplied annotation filter, which filters out annotations from the `java.lang` and `org.springframework.lang` packages.

Since none of the annotations from the `java.lang` and `org.springframework.lang` packages is _repeatable_, the filtering of repeatable annotations is unnecessary.

In light of that, I am changing this from a "bug" to a "task" issue in order to remove the method from the package-private API and to remove the filtering of repeatable annotations altogether.

**Comment by sbrannen** [comment]
> It turns out that `AnnotationTypeMappings.forAnnotationType(Class<? extends Annotation>, AnnotationFilter)` is not invoked outside of the `AnnotationTypeMappings` class itself.

Actually, I made a mistake while analyzing this in my IDE. An `AnnotationFilter` can be supplied to `AnnotationTypeMappings` via the `from` factory methods in `TypeMappedAnnotations` via the `from` factory methods in `MergedAnnotations` which are part of the public API.

Thus, we will merge this PR as a bug fix.

File: spring-core/src/main/java/org/springframework/core/annotation/AnnotationTypeMappings.java (hunk 1 above)
Comment (philwebb) [reply spring-core/src/main/java/org/springframework/core/annotation/AnnotationTypeMappings.java:94]: I think it was a typo. A little concerning that no test covered it.

**Comment by sbrannen** [comment]
This has been merged into `5.2.x` and `master`.

Thanks

**Pull request closed without merging**